include/cgimap/http.hpp

```diff
@@ -115,17 +116,32 @@ class gone : public exception {
 };
 
 /**
+ * Indicates that the client is not authorized to perform the request.
+ */
+class unauthorized : public exception {
+public:
+  unauthorized(const std::string &message);
+};
+
+/**
  * Decodes a url-encoded string.
  */
 std::string urldecode(const std::string &s);
 
 /**
+ * Encodes a string to canonical url-encoding.
+ *
+ * This is compatible with the OAuth 1.0a definition of url-encoding.
+ */
+std::string urlencode(const std::string &s);
+
+/**
  * Parses a query string into a key-value map.
  *
  * The string should already have been url-decoded (i.e: no %-encoded
  * chars remain).
  */
-std::map<std::string, std::string> parse_params(const std::string &p);
+std::vector<std::pair<std::string, std::string> > parse_params(const std::string &p);
```

Comment:
That was to handle duplicates in commit 4dfa54cdb42eb48b3258b98aa4421d255acf08f6. One of the [OAuth examples](https://tools.ietf.org/html/rfc5849#section-3.4.1.3.1) has duplicates coming from a mixture of GET query and POST form-encoded parameters, and the code wasn't handling that correctly. It seemed easier to remove the assumption that query parameters had unique keys throughout the code than to convert back and forth between data structures.

As for `multimap` vs `vector<pair>`... I just figured the latter was simpler to deal with, but perhaps it would be a nice code cleanup later to move to `multimap`.